Snowflake.Data.Tests/UnitTests/SFS3ClientTest.cs

```diff
@@ -0,0 +1,394 @@
+
+        // Mock upload file size
+        const int UploadFileSize = 9999;
+
+        // The mock client and metadata
+        SFS3Client _client;
+        SFFileMetadata _fileMetadata;
+        AmazonS3Config _clientConfig;
+
+        [SetUp]
+        public void BeforeTest()
+        {
+            _fileMetadata = new SFFileMetadata()
+            {
+                stageInfo = new PutGetStageInfo()
+                {
+                    endPoint = Endpoint,
+                    isClientSideEncrypted = true,
+                    location = Location,
+                    locationType = SFRemoteStorageUtil.S3_FS,
+                    path = LocationPath,
+                    presignedUrl = null,
+                    region = Region,
+                    stageCredentials = _stageCredentials,
+                    storageAccount = null
+                }
+            };
+
+            _clientConfig = new AmazonS3Config();
+            _clientConfig.RegionEndpoint = RegionEndpoint.GetBySystemName(_fileMetadata.stageInfo.region);
+
+            _cancellationToken = new CancellationToken();
+        }
+        [Test]
+        [Ignore("S3ClientTest")]
+        public void S3ClientTestDone()
+        {
+            // Do nothing;
+        }
+
+        [Test]
+        public void TestExtractBucketNameAndPath()
+        {
+            // Arrange
+            var mockAmazonS3Client = new Mock<AmazonS3Client>(AwsKeyId, AwsSecretKey, AwsToken, _clientConfig);
+            _client = new SFS3Client(_fileMetadata.stageInfo, MaxRetry, Parallel, _proxyCredentials, mockAmazonS3Client.Object);
+            _fileMetadata.client = _client;
+            // Split LOCATION based on the first '/' character
+            string[] bucketAndKey = Location.Split(new[] { '/' }, 2);
+
+            // Act
+            var location = _client.ExtractBucketNameAndPath(_fileMetadata.stageInfo.location);
+
+            // Assert
+            Assert.AreEqual(bucketAndKey[0], location.bucket);
+            Assert.AreEqual(bucketAndKey[1], location.key);
+        }
+
+        [Test]
+        [TestCase(MockS3Client.AwsStatusOk, ResultStatus.UPLOADED)]
+        [TestCase(SFS3Client.EXPIRED_TOKEN, ResultStatus.RENEW_TOKEN)]
+        [TestCase(SFS3Client.NO_SUCH_KEY, ResultStatus.NOT_FOUND_FILE)]
+        [TestCase(MockS3Client.AwsStatusError, ResultStatus.ERROR)] // Any error that isn't the above will return ResultStatus.ERROR
+        public void TestGetFileHeader(string requestKey, ResultStatus expectedResultStatus)
+        {
+            // Arrange
+            var mockAmazonS3Client = new Mock<AmazonS3Client>(AwsKeyId, AwsSecretKey, AwsToken, _clientConfig);
+            mockAmazonS3Client.Setup(client => client.GetObjectAsync(It.IsAny<GetObjectRequest>(), It.IsAny<CancellationToken>()))
+                .Returns<GetObjectRequest, CancellationToken>((request, cancellationToken) =>
+                {
+                    if (request.BucketName == HttpStatusCode.OK.ToString())
```

Comment:
response is same for sync and async (difference about wrapping it in Task) so consider extracting them

Reply:
Ok, extracted response into its own method